src/audio/tdfb/tdfb_direction.c

```diff
@@ -0,0 +1,497 @@
+#include <sof/math/trig.h>
+#include <user/eq.h>
+#include <stdint.h>
+#include "tdfb_math.h"
+
+#ifdef DEBUG
+#include <stdio.h>
+FILE *fh;
+#endif
+
+/* Generic definitions */
+#define ZERO_POINT_FIVE		Q_CONVERT_FLOAT(0.5, 30)
+#define SPEED_OF_SOUND		340				/* m/s Q0 */
+#define COEF_DEG_TO_RAD		Q_CONVERT_FLOAT(0.017453, 15)	/* Q1.15 */
+#define COEF_RAD_TO_DEG		Q_CONVERT_FLOAT(57.296, 9)	/* Q6.9 */
+#define PI_Q12			Q_CONVERT_FLOAT(3.1416, 12)
+#define PIMUL2_Q12		Q_CONVERT_FLOAT(6.2832, 12)
+#define PIDIV2_Q12		Q_CONVERT_FLOAT(1.5708, 12)
+
+/* Sound levels filtering related, these form a primitive voice activity
+ * detector. Sound levels below ambient estimate times threshold (kind of dB offset)
+ * are not scanned for sound direction.
+ */
+#define SLOW_LEVEL_C1		Q_CONVERT_FLOAT(0.001, 30)
+#define SLOW_LEVEL_C2		Q_CONVERT_FLOAT(1 - 0.001, 30)
+#define FAST_LEVEL_C1		Q_CONVERT_FLOAT(0.5, 30)
+#define FAST_LEVEL_C2		Q_CONVERT_FLOAT(1 - 0.5, 30)
+#define LEVEL_THRESHOLD		Q_CONVERT_FLOAT(2.5, 12)
+
+/* Iteration parameters, smaller step and larger iterations is more accurate but
+ * consumes more cycles.
+ */
+#define AZ_STEP			Q_CONVERT_FLOAT(0.3927, 12)	/* PI/16 as Q4.12 */
+#define AZ_ITERATIONS		5				/* loops in min err search */
+#define SOURCE_DISTANCE		Q_CONVERT_FLOAT(5.0, 12)	/* source distance in m Q4.12 */
+
+/* Sound direction angle filtering */
+#define SLOW_AZ_C1		Q_CONVERT_FLOAT(0.01, 15)
+#define SLOW_AZ_C2		Q_CONVERT_FLOAT(1 - 0.01, 15)
+
+/* Threshold for notifying user space, no more often than every 100 ms */
+#define CONTROL_UPDATE_MIN_TIME	Q_CONVERT_FLOAT(100e-3, 16)
+
+/* Emphasis filter for sound direction */
+uint32_t iir_emphasis_48k[13] = {
+	0x00000001, 0x00000001, 0x00000000, 0x00000000,
+	0x00000000, 0x00000000, 0xd7b49168, 0x62c27f22,
+	0x19712b75, 0xcd1da916, 0x19712b75, 0x00000000,
+	0x00007fb2
+};
+
+uint32_t iir_emphasis_16k[13] = {
+	0x00000001, 0x00000001, 0x00000000, 0x00000000,
+	0x00000000, 0x00000000, 0xef646e59, 0x2cc88203,
+	0x0fb60984, 0xe093ecf7, 0x0fb60984, 0x00000000,
+	0x00007fb2
+};
+
+/* Function to copy data from component source to cross correlation buffer with emphasis
+ * filtering. Skip if not tracking.
+ */
+void tdfb_direction_copy_emphasis(struct tdfb_comp_data *cd, int channels, int *channel, int32_t x)
+{
+	if (!cd->direction_updates)
+		return;
+
+	*cd->direction.wp = iir_df2t(&cd->direction.emphasis[*channel], x) >> 16;
+	(*channel)++;
+	cd->direction.wp++;
+	tdfb_cinc_s16(&cd->direction.wp, cd->direction.d_end, cd->direction.d_size);
+	if (*channel == channels)
+		*channel = 0;
+}
+
+static inline int16_t max_mic_distance(struct tdfb_comp_data *cd)
+{
+	int16_t dx;
+	int16_t dy;
+	int16_t dz;
+	int32_t d2;
+	int32_t d2_max = 0;
+	int i;
+	int j;
+
+	/* Max lag based on largest array dimension. Microphone coordinates are Q4.12 meters */
+#ifdef DEBUG
+	for (i = 0; i < cd->config->num_mic_locations; i++) {
+		printf("printf: Mic%d (%d, %d, %d)\n", i, cd->mic_locations[i].x,
```

Comment:
we can use trace for this, testbench will show it too.

Reply:
Yep, I've changed those. Test scripts now parse trace output, though I will submit that PR a bit later since there's still cleanup need and other dependencies.